### ISSUE TITLE
Support large files

### ISSUE THREAD

**Issue by genthalili** [mariodavid/cuba-component-data-import#86]
When I say large file, I mean 40K lines (1MB excel file).

I tested the add-on with larger files and here is my feedback : 
Today the import simple and association works well (with a good configuration) for 100 lines of imports, but this becomes way longue for files of 40K lines.
I see 2 issues : 
1. Importing "large" files (simple) should use bulk upload (easy part) every 1000 record bulk commit etc..
2. Importing "large" files (association) it's the most complicated one! it should do [40K lookups] * [number of association columns] * memory storage

My suggestion : 
Use temporary table for associations or (SQL query building list of entities) to create the joins in DDB.
and commit by buckets (1000 max).

@mariodavid What do you think?

**Comment by mariodavid** [comment]
hi,

regarding 1: might be another option. But generally I would let it up to the user to decide. So perhaps it becomes another `ImportTransactionStrategy`.

regarding 2: the situation is much worse depending on how much unique configurations are configured... 

Generally I think the import does not need to be fast, because it will not be done every hour or so, but rather it is a one off thing. However: it should also not bring the server down because of the memory. The problem is that this really depends on the server you are running your application on & the size of the file. For now I think the data can be splitted into multiple files with something like this: http://filesplit.org/. Perhaps in the future there is time to look at speed improvements.

**Comment by genthalili** [comment]
Hi, 

An other `ImportTransactionStrategy` would be the solution, I'm not sure if `entityImportExportAPI` supports this, but what I have been using on my previous projects (code below) and it's working just fine, even with million entities.
`````
        Transaction tx = persistence.createTransaction();
        try {
            EntityManager entityManager = persistence.getEntityManager();
            int i = 0;
            for (Entity t : entities) {
                entityManager.persist(t);
                i++;
                if (i % batchSize == 0) {
                    tx.commitRetaining();
                    entityManager = persistence.getEntityManager();
                }
            }
            tx.commit();
        } finally {
            tx.close();
        }
``````

For the second issue: Maybe you are not woking with large tables but this is the case for us. We have almost 100K rows in each table and most of them need to be rieched by foreign keys with any key as input (because I use UUID and they are difficult to remember :) ). So for me a batch lookup solution is very important, and I think the powerful of this add-on is on associations.

I have bee looking for a easy solution, but haven't found yet, if you have any, please let me know.

**Comment by mariodavid** [comment]
Hi,

you are right that there might be a need to support big files. `EntityImportExportAPI` will do the import in a single transaction. But thats not a real problem, it is possible to go around that (as I did with the `ImportTransactionStrategy.TRANSACTION_PER_ENTITY` where I basically called `EntityImportExportAPI` for every entity. So from an import perspective it is totally achievable to create another mode for batch.

But currently the way the wizard works is not at all optimized for large files. There are several features, that play against working with large files. E.g. the whole file is parsed in the FE server (which means all data rows will be in memory. The reason is that there is this preview feature. But this does not make sense anymore when dealing with large files, because it is not a manual process anymore, because the amount of data will just be too large. Also #28 and #29 will work against the idea of big files.

So a proper solution to this problem is probably that:
1. the `GenericImportService` should accept another programmatic way of interacting with it: passing in a FileDescriptor so the parsing can happen is batches
2. the `GenericImportService` should allow to operate in a batch mode with a new ImportStrategy
3. there might be another wizard option so operate in "large file mode" which means deactivating different features

However, I don't know if this improvements will eliminate the bottleneck, because I don't know exactly what the bottleneck is. If You configure 10 unique configurations, it does not really matter if you do it in batch mode or not, because perhaps the amount of memory spend for parsing the file is not really the biggest problem...

Perhaps I will implement something in the future (at least 1. & 2.) to make that feature accessible programmatically. Then it only depends on the configuration that will be used for the import (especially unique configurations and custom groovy scripts).

**Comment by mariodavid** [comment]
hi @genthalili ,

i just released version 0.9.0 which contains the `DataImportAPI`. With that it will remove the burden, that the import wizard has to be used. Therefore it _can_ be faster.

I would be great if you can test it out. I will close this issue for now.
